### PR TITLE
[Sparse] GAT Example

### PR DESCRIPTION
**Pull request by mufeili**
## Description
Test accuracy numbers I got: 0.77 ~ 0.82

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [x] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [ ] All changes have test coverage
- [x] Code is well-documented
- [x] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: examples/sparse/gat.py [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+        nn.init.xavier_normal_(self.a_l, gain=gain)
+        nn.init.xavier_normal_(self.a_r, gain=gain)
+
+    def forward(self, A_hat, Z):
+        Z = self.dropout(Z)
+        Z = self.W(Z).view(Z.shape[0], self.out_size, self.num_heads)
+        # a^T [Wh_i || Wh_j] = a_l Wh_i + a_r Wh_j
+        e_l = (Z * self.a_l).sum(dim=1)
+        e_r = (Z * self.a_r).sum(dim=1)
+        e = e_l[A_hat.row] + e_r[A_hat.col]
```
Comment (czkkkkkk) [inline examples/sparse/gat.py:39]: After we migrate to the new sparse library, we need to check whether these examples can work and may refactor some of the interfaces. Could you file it as an issue?

**Review by czkkkkkk — APPROVED** [review]

File: examples/sparse/gat.py (hunk 1 above)
Comment (mufeili) [reply examples/sparse/gat.py:39]: I think this applies to all examples. Perhaps we can discuss it further during the weekly meeting.

File: examples/sparse/gat.py [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+        A_hat.val = F.leaky_relu(e)
+        A_atten = A_hat.softmax()
+        A_atten.val = self.dropout(A_atten.val)
+        return bspmm(A_atten, Z)
```
Comment (frozenbugs) [inline examples/sparse/gat.py:43]: Keep in mind that this is an example for showcase sparse API as in sparse directory.
Always highlight the piece that takes advantage of sparse API.

File: examples/sparse/gat.py [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+    def reset_parameters(self):
+        gain = nn.init.calculate_gain("relu")
+        nn.init.xavier_normal_(self.W.weight, gain=gain)
+        nn.init.xavier_normal_(self.a_l, gain=gain)
+        nn.init.xavier_normal_(self.a_r, gain=gain)
+
+    def forward(self, A_hat, Z):
+        Z = self.dropout(Z)
+        Z = self.W(Z).view(Z.shape[0], self.out_size, self.num_heads)
+        # a^T [Wh_i || Wh_j] = a_l Wh_i + a_r Wh_j
```
Comment (frozenbugs) [inline examples/sparse/gat.py:36]: nit: which lines implement this formula? L37-L39? Add blank lines before and after the code block.

**Review by frozenbugs — APPROVED** [review]

File: examples/sparse/gat.py (hunk 1 above)
Comment (frozenbugs) [reply examples/sparse/gat.py:39]: Yes, we will update all examples and enable CI after we migrate to the new sparse library. Using https://github.com/dmlc/dgl/issues/4948 to track.

File: examples/sparse/gat.py [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+    def forward(self, A_hat, Z):
+        Z = self.dropout(Z)
+        Z = self.W(Z).view(Z.shape[0], self.out_size, self.num_heads)
+
+        # a^T [Wh_i || Wh_j] = a_l Wh_i + a_r Wh_j
+        e_l = (Z * self.a_l).sum(dim=1)
+        e_r = (Z * self.a_r).sum(dim=1)
+        e = e_l[A_hat.row] + e_r[A_hat.col]
+
+        A_hat.val = F.leaky_relu(e)
```
Comment (jermainewang) [inline examples/sparse/gat.py:46]: Still not like this, although L48 is fine for me because `A_atten` is a different matrix than `A_hat`.

**Pull request merged**